### PR TITLE
Implement create card service

### PR DESCRIPTION
**Pull request by cjsewell**
Implement the create card service for gateways which support it.

Tested with DPS PxPay

Relates to #158

### CODE REVIEW

**Comment by bummzack** [comment]
Looking good. Could you provide Unit-Tests with this?

**Comment by cjsewell** [comment]
Tests added. Im reasonably sure they are implemented correctly. 
Let me know if not

**Comment by cjsewell** [comment]
@bummzack  It appears my tests have broken something. Are you able to assist it what's gone wrong?

**Comment by bummzack** [comment]
I'm gonna have a look. What's certainly *not* a good idea is to create a custom Gateway for tests. I strongly advise to use a Mock gateway for that case.

File: tests/payment.yml [hunk 1]
```diff
@@ -69,3 +77,7 @@ GatewayMessage:
     ClassName: AuthorizedResponse
     Reference: authorizedPaymentReceipt
     PaymentID: =>Payment.payment6
+  message6:
```
Comment (bummzack) [inline tests/payment.yml]: Should have it's own number, currently there are two entries named `message6`. This throws off the other tests.

**Comment by bummzack** [comment]
I've given this a closer look. I'm wondering why you chose to mimic the behavior of `Authorize` and `Capture`. Can you explain how one would normally use *create card*? I thought it's an API call, similar to `Refund` or `Void`, where there's no user-interaction. Or is it like `Purchase`, where you might get redirected to an offsite form?

**Comment by cjsewell** [comment]
Its similar to Purchase, where you get redirected. 
For DPS at least, it is the same as purchase, just with an additional flag.

**Comment by bummzack** [comment]
Ok, thanks for clarifying. So what's needed to make tests working is: Fix the fixture and revert/remove all the `SS_Dummy` stuff. Then use a mock gateway that supports "createCard". Adding this to `CreateCardServiceTest` should do the trick:

```php
public function testDummyOnSitePayment()
{
    $stubGateway = $this->buildDummyGatewayMock(true);
    $payment = $this->payment->setGateway('Dummy');
    $payment->setFailureUrl('my/cancel/url')->setSuccessUrl('my/return/url');

    $service = $this->getService($payment);
    $mockFactory = $this->stubGatewayFactory($stubGateway);
    $service->setGatewayFactory($mockFactory);
    Injector::inst()->registerService($mockFactory, 'Omnipay\Common\GatewayFactory');

    parent::testDummyOnSitePayment();
}

public function testFailedDummyOnSitePayment()
{
    $stubGateway = $this->buildDummyGatewayMock(false);
    $payment = $this->payment->setGateway('Dummy');
    $payment->setFailureUrl('my/cancel/url')->setSuccessUrl('my/return/url');

    $service = $this->getService($payment);
    $mockFactory = $this->stubGatewayFactory($stubGateway);
    $service->setGatewayFactory($mockFactory);
    Injector::inst()->registerService($mockFactory, 'Omnipay\Common\GatewayFactory');

    parent::testFailedDummyOnSitePayment();
}

protected function buildDummyGatewayMock($successValue)
{
    //--------------------------------------------------------------------------------------------------------------
    // Payment request and response

    $mockPaymentResponse = $this->getMockBuilder('Omnipay\Dummy\Message\Response')
        ->disableOriginalConstructor()->getMock();

    $mockPaymentResponse->expects($this->any())
        ->method('isSuccessful')->will($this->returnValue($successValue));

    $mockPaymentRequest = $this->getMockBuilder('Omnipay\Dummy\Message\AuthorizeRequest')
        ->disableOriginalConstructor()->getMock();

    $mockPaymentRequest->expects($this->any())->method('send')->will($this->returnValue($mockPaymentResponse));

    //--------------------------------------------------------------------------------------------------------------
    // Build the gateway

    $stubGateway = $this->getMockBuilder('Omnipay\Common\AbstractGateway')
        ->setMethods(array('createCard', 'getName'))
        ->getMock();

    $stubGateway->expects($this->once())
        ->method('createCard')
        ->will($this->returnValue($mockPaymentRequest));

    return $stubGateway;
}
```

These tests really assume that `createCard` is identical to other actions, like `authorize`. Not sure if there are other gateways that handle this stuff differently. If your code works with DPS, it's a start though.

**Comment by cjsewell** [comment]
Ahh I see, that's much cleaner.

Updated and all tests appear to pass.

Thanks for your help

**Comment by bummzack** [comment]
Thanks a lot for you work!

**Pull request merged**